Source/CarthageKit/Project.swift

```diff
@@ -13,23 +13,56 @@ import ReactiveCocoa
 /// Carthage’s bundle identifier.
 public let CarthageKitBundleIdentifier = NSBundle(forClass: Project.self).bundleIdentifier!
 
+/// The fallback dependencies URL to be used in case
+/// the intended ~/Library/Caches/org.carthage.CarthageKit cannot
+/// be found or created.
+private let fallbackDependenciesURL: NSURL = {
+	let homePath: String
+	if let homeEnvValue = NSProcessInfo.processInfo().environment["HOME"] as? NSString {
+		homePath = homeEnvValue.stringByAppendingPathComponent(".carthage")
+	} else {
+		homePath = "~/.carthage".stringByExpandingTildeInPath
+	}
+	return NSURL.fileURLWithPath(homePath, isDirectory:true)!
+}()
+
 /// ~/Library/Caches/org.carthage.CarthageKit/
 private let CarthageUserCachesURL: NSURL = {
-	let URL: Result<NSURL, NSError> = try({ (error: NSErrorPointer) -> NSURL? in
-		NSFileManager.defaultManager().URLForDirectory(NSSearchPathDirectory.CachesDirectory, inDomain: NSSearchPathDomainMask.UserDomainMask, appropriateForURL: nil, create: true, error: error)
-	})
-
-	let fallbackDependenciesURL = NSURL.fileURLWithPath("~/.carthage".stringByExpandingTildeInPath, isDirectory:true)!
+	let fileManager = NSFileManager.defaultManager()
+	
+	let URLResult: Result<NSURL, NSError> = try({ (error: NSErrorPointer) -> NSURL? in
+		fileManager.URLForDirectory(NSSearchPathDirectory.CachesDirectory, inDomain: NSSearchPathDomainMask.UserDomainMask, appropriateForURL: nil, create: true, error: error)
+	}).flatMap { cachesURL in
+		let dependenciesURL = cachesURL.URLByAppendingPathComponent(CarthageKitBundleIdentifier, isDirectory: true)
+		let dependenciesPath = dependenciesURL.absoluteString!
+		
+		let result: Result<NSURL, NSError>
+		if fileManager.fileExistsAtPath(dependenciesPath, isDirectory:nil) {
+			if fileManager.isWritableFileAtPath(dependenciesPath) {
+				result = Result(value: dependenciesURL)
+			} else {
+				let error = NSError(domain: CarthageKitBundleIdentifier, code: 0, userInfo: nil)
+				result = Result(error: error)
+			}
+		} else {
+			result = try { (error: NSErrorPointer) -> NSURL? in
+				let attributes = [NSFilePosixPermissions : NSNumber(short:755)]
+				fileManager.createDirectoryAtURL(dependenciesURL, withIntermediateDirectories: true, attributes:attributes, error: error)
```

Comment:
I think we should be checking the return value here and creating an error if it fails.